doc: Add design doc for thin-snapshots (#446)
topolvm/topolvm: docs/proposals/thin-snapshots-restore.md

## Flow of Operations for Snapshot Creation

* An user creates a `VolumeSnapshot` custom resource.
* The `snapshot-controller` which was watching the `VolumeSnapshot` object, sends a `Create VolumesnapshotContent` req to the api-server.
* The `external-snapshotter` kubernetes sidecar receives this request, and then sends a `CreateSnapshot` gRPC call to TopoLVM CSI driver.
* This `CreateSnapshot` request is received by the `topolovm-controller` which then create a `LogicalVolume` CR.
* `topolvm-node` on the target node finds the earlier created `LogicalVolume` CR.
* `topolvm-node` sends a volume create request to lvmd to create a thin-snapshot from the thin-volume datasource.
* `lvmd` creates an LVM logical volume as requested.
* `topolvm-node` updates the status of LogicalVolume.
* `topolvm-controller` finds the updated status of LogicalVolume.
* `topolvm-controller` sends the success (or failure) to `external-snapshotter` sidecar.

## Proposed LogicalVolume CR for Snapshots

```yaml
apiVersion: topolvm.cybozu.com/v1
kind: LogicalVolume
metadata:
  name: snapcontent-b083470e-8293-47cc-810d-9561bd1754e6 # restore pv name
  annotations:

> spec:

Author: Nice catch! that was an overlook from my side, updated now :+1:

```

## LV Operations

* Snapshot creation from a thinly-provisioned volume:
    ```bash
    $ lvcreate -s --name thinsnap VG0/thinvolume
    ```

* Activation of a logical volume

  A thinly-provisioned snapshot will be activated as Read-Only for snapshots and Read-Write for Restore scenarios.

  For e.g: Activating a logical volume for Read-Write operations:

  ```bash
  $ lvchange -ay -K VG0/thinsnap
  ```

* Restoring a Snapshot

  ```bash
  $ lvcreate -s --name thin-restored VG0/thinsnap
  ```

## Deployment Changes

* Add `external-snapshotter` sidecar.

* Perform `snapshot-controller` deployment.

## Metrics and Monitoring

Should be already covered during adding support for `thin` logical volumes. See [thin-provisioning proposal doc](https://github.com/topolvm/topolvm/pull/442).

## Additional Notes

* When creating a thin snapshot volume, you do not specify the size of the volume.

* Thin snapshot volumes are not supported across the nodes in a cluster. The snapshot volume must be exclusively activated on only one cluster node.
## Open Questions

- Should we only allow thick snapshot creation for thick volumes and thin-snapshots for thinly-provsioned volumes?

>   - If yes, then we don't need to specify `type` parameter for snapshot volumes in the LogicalVolume CR; as we can assume thin-snapshots if the parent is a thinly-provisioned volume.

Author: Sure, we can keep the `type` parameter

# Snapshot & Restore

> > Note: Based on thin-provisioning of volumes. Please refer to [thin-provisioning proposal doc](https://github.com/topolvm/topolvm/pull/442).

Reviewer: Please correct this as well.
```suggestion
> Note: Based on thin-provisioning of volumes. Please refer to [thin-provisioning proposal doc](docs/proposals/thin-volumes.md).
```

Author: Great catch! Updated :+1:

* Blocks common to recursive snapshots are also shared in the thin pool. There is no limit to or degradation from sequences of snapshots.

* To provide a restored PVC to the user, we create another thin-provisioned snapshot of the available snapshot; and since lvm snapshots,  by design, are writable, so they can be provided to users to use as a volume for a new PVC.

* This way, we also don't lose the identity of the snapshot after restoring; so we can restore a snapshot multiple times, to create multiple copies.

* These thin snapshot volumes are activated as `Read-Write` copies of the parent, and provided to kubernetes users as a PersistentVolume object.

![](https://i.imgur.com/bwsK1Dl.png)

> Note: The Snapshots, Restored volumes and Clones are all independent of their parent deletion.

## Flow of Operations for Snapshot Creation

* An user creates a `VolumeSnapshot` custom resource.
* The `snapshot-controller` which was watching the `VolumeSnapshot` object, sends a `Create VolumesnapshotContent` req to the api-server.
* The `external-snapshotter` kubernetes sidecar receives this request, and then sends a `CreateSnapshot` gRPC call to TopoLVM CSI driver.

> * This `CreateSnapshot` request is received by the `topolovm-controller` which then create a `LogicalVolume` CR.

Author: Nice catch! Updated now :+1:

    ```

* Activation of a logical volume

  A thinly-provisioned snapshot will be activated as Read-Only for snapshots and Read-Write for Restore scenarios.

  For e.g: Activating a logical volume for Read-Write operations:

  ```bash
  $ lvchange -ay -K VG0/thinsnap
  ```

* Restoring a Snapshot

  ```bash
  $ lvcreate -s --name thin-restored VG0/thinsnap
  ```

## VolumeSnapshotClass and VolumeSnapshot

### VolumeSnapshotClass

```yaml
apiVersion: snapshot.storage.k8s.io/v1
kind: VolumeSnapshotClass
metadata:
  name: topolvm-snapclass
driver: topolvm.cybozu.com
deletionPolicy: Delete
parameters:
  # This field is for selecting the mode of restoring volume. It must be `local`
  # for snapshots intended to be created on the same node.
  "topolvm.cybozu.com/restore-mode": "local"
```

### VolumeSnapshot

```yaml
apiVersion: snapshot.storage.k8s.io/v1
kind: VolumeSnapshot
metadata:
  name: new-snapshot
spec:
  volumeSnapshotClassName: topolvm-snapclass
  source:
    persistentVolumeClaimName: snapshot-pvc
```

## Deployment Changes

* Add `external-snapshotter` sidecar.

* Perform `snapshot-controller` deployment.

## Metrics and Monitoring

> Should be already covered during adding support for `thin` logical volumes. See [thin-provisioning proposal doc](https://github.com/topolvm/topolvm/pull/442).

Reviewer: PR #442 has been merged, please change the URL.
```suggestion
Should be already covered during adding support for `thin` logical volumes. See [thin-provisioning proposal doc](docs/proposals/thin-volumes.md).
```

Author: Thanks, updated :+1:

# Snapshot & Restore

> > Note: Based on thin-provisioning of volumes. Please refer to [thin-provisioning proposal doc](thin-volumes.md).

Author: Added a warning at the beginning of the doc and a limitation section

## Goals

Allow users to create thin-snapshots from thinly-provisioned logical volumes.
These thin-snapshots, then set the foundation for allowing creation of snapshots of persistent volume claims.

* **Allow Snapshot Creation**: Enable users to take snapshots of PVs that are provisioned by TopoLVM.

* **Restoring Snapshots**: Based on a volumesnapshot, a new PVC can be created that contains the contents of the snapshot.

## Introduction

> 

Reviewer 2: Please add  a section for Limitations.

* This way, we also don't lose the identity of the snapshot after restoring; so we can restore a snapshot multiple times, to create multiple copies.

* These thin snapshot volumes are activated as `Read-Write` copies of the parent, and provided to kubernetes users as a PersistentVolume object.

![](https://i.imgur.com/bwsK1Dl.png)

> Note: The Snapshots, Restored volumes and Clones are all independent of their parent deletion.

## Flow of Operations for Snapshot Creation

* A user creates a `VolumeSnapshot` custom resource.
* The `snapshot-controller` which was watching the `VolumeSnapshot` object, sends a `Create VolumesnapshotContent` req to the api-server.
* The `external-snapshotter` kubernetes sidecar receives this request, and then sends a `CreateSnapshot` gRPC call to TopoLVM CSI driver.
* This `CreateSnapshot` request is received by the `topolvm-controller` which then create a `LogicalVolume` CR.
* `topolvm-node` on the target node finds the earlier created `LogicalVolume` CR.

> * `topolvm-node` sends a volume create request to lvmd to create a thin-snapshot from the thin-volume datasource.

Author: I believe we can leave this one as is as it already mentions ` from the thin-volume datasource`.

    $ lvcreate -s --name thinsnap VG0/thinvolume
    ```

* Activation of a logical volume

  A thinly-provisioned snapshot will be activated as Read-Only for snapshots and Read-Write for Restore scenarios.

  For e.g: Activating a logical volume for Read-Write operations:

  ```bash
  $ lvchange -ay -K VG0/thinsnap
  ```

* Restoring a Snapshot

  ```bash
  $ lvcreate -s --name thin-restored VG0/thinsnap
  ```

## VolumeSnapshotClass and VolumeSnapshot

### VolumeSnapshotClass

```yaml
apiVersion: snapshot.storage.k8s.io/v1
kind: VolumeSnapshotClass
metadata:
  name: topolvm-snapclass
driver: topolvm.cybozu.com
deletionPolicy: Delete
```

### VolumeSnapshot

```yaml
apiVersion: snapshot.storage.k8s.io/v1
kind: VolumeSnapshot
metadata:
  name: new-snapshot
spec:
  volumeSnapshotClassName: topolvm-snapclass
  source:
    persistentVolumeClaimName: snapshot-pvc
```

## Restoring the Snapshot to a new PVC

When creating a snapshot or restoring it to a new volume; both the snapshot and the restored volume need to be provisioned
on the same node as the source volume.

Since the a snapshots can only be taken of a BOUND PVC, and volumeBinding is `waitforfirstconsumer`; it is safe to

> assume that the source PVC already has an application POD using it. So while creating a restored PVC from the snapshot of source, we need

Author: Thanks for the suggestion! Updated to use Node selectors

> to schedule the pod/app on the same node. To do that, we can make use of pod affinity.

Author: Thanks for the suggestion. True, although POD Affinity works is convenient if the source PVC has an application using it, but Node Affinity covers all the use cases.
Updated to use node affinity now

  For e.g: Activating a logical volume for Read-Write operations:

  ```bash
  $ lvchange -ay -K VG0/thinsnap
  ```

* Restoring a Snapshot

  ```bash
  $ lvcreate -s --name thin-restored VG0/thinsnap
  ```

## VolumeSnapshotClass and VolumeSnapshot

### VolumeSnapshotClass

```yaml
apiVersion: snapshot.storage.k8s.io/v1
kind: VolumeSnapshotClass
metadata:
  name: topolvm-snapclass
driver: topolvm.cybozu.com
deletionPolicy: Delete
```

### VolumeSnapshot

```yaml
apiVersion: snapshot.storage.k8s.io/v1
kind: VolumeSnapshot
metadata:
  name: new-snapshot
spec:
  volumeSnapshotClassName: topolvm-snapclass
  source:
    persistentVolumeClaimName: snapshot-pvc
```

## Restoring the Snapshot to a new PVC

When creating a snapshot or restoring it to a new volume; both the snapshot and the restored volume need to be provisioned
on the same node as the source volume.

Since the a snapshots can only be taken of a BOUND PVC, and volumeBinding is `waitforfirstconsumer`; it is safe to
assume that the source PVC already has an application POD using it. So while creating a restored PVC from the snapshot of source, we need

> to schedule the pod/app on the same node. To do that, we can make use of pod affinity.

Author: Updated :+1:

In the `pod-restore.yaml`, we must schedule the restore pod on the same node as the source by using the source pod name as the label for pod affinity.
Here, before creating the pod, an user just has to mention the name of the source pvc pod.

For example:

Source application pod uses the label
`[app.kubernetes.io/name](http://app.kubernetes.io/name): my-pod`.

To schedule the restored pod on the same node,
the restored pod has the pod affinity set for the source PVC label.

`pod.yaml`

```yaml
apiVersion: v1
kind: Pod
metadata:
  name: my-pod
  labels:
    [app.kubernetes.io/name](http://app.kubernetes.io/name): my-pod
spec:
  containers:
...
...
...
```

`pod-restore.yaml`

```yaml
apiVersion: v1
kind: Pod
metadata:
  name: pod-restore
spec:
# The Restored PVC needs to be provisioned on the same node as the source volume.
  affinity:
    podAffinity:
      requiredDuringSchedulingIgnoredDuringExecution:

>     * labelSelector:

Reviewer: ```suggestion
      - labelSelector:
```

          matchExpressions:

>         * key:  [app.kubernetes.io/name](http://app.kubernetes.io/name)

Reviewer: ```suggestion
            - key: app.kubernetes.io/name
```

```

### VolumeSnapshot

```yaml
apiVersion: snapshot.storage.k8s.io/v1
kind: VolumeSnapshot
metadata:
  name: new-snapshot
spec:
  volumeSnapshotClassName: topolvm-snapclass
  source:
    persistentVolumeClaimName: snapshot-pvc
```

## Restoring the Snapshot to a new PVC

When creating a snapshot or restoring it to a new volume; both the snapshot and the restored volume need to be provisioned
on the same node as the source volume.

Since the a snapshots can only be taken of a BOUND PVC, and volumeBinding is `waitforfirstconsumer`; it is safe to
assume that the source PVC already has an application POD using it. So while creating a restored PVC from the snapshot of source, we need
to schedule the pod/app on the same node. To do that, we can make use of pod affinity.

In the `pod-restore.yaml`, we can schedule the restore pod on the same node as the source using the source pod name as the label for pod affinity.
Here, before creating the pod, an user just has to mention the name of the source pvc pod.

For example:

Source application pod uses the label
`[app.kubernetes.io/name](http://app.kubernetes.io/name): my-pod`.

To schedule the restored pod on the same node,

> the restored pod has the pod affinity set for te source PVC label.

Reviewer: typo
```suggestion
the restored pod has the pod affinity set for the source PVC label.
```

Author: Great catch! Updated :+1:
